src/Datadog.AutoInstrumentation.NativeLoader/cor_profiler.cpp

```diff
@@ -0,0 +1,4587 @@
+            riid == __uuidof(ICorProfilerCallback2) || riid == __uuidof(ICorProfilerCallback) || riid == IID_IUnknown)
+        {
+            *ppvObject = this;
+            this->AddRef();
+            return S_OK;
+        }
+
+        *ppvObject = nullptr;
+        return E_NOINTERFACE;
+    }
+
+    ULONG STDMETHODCALLTYPE CorProfiler::AddRef(void)
+    {
+        Debug("CorProfiler::AddRef");
+        return std::atomic_fetch_add(&this->m_refCount, 1) + 1;
+    }
+
+    ULONG STDMETHODCALLTYPE CorProfiler::Release(void)
+    {
+        Debug("CorProfiler::Release");
+        int count = std::atomic_fetch_sub(&this->m_refCount, 1) - 1;
+
+        if (count <= 0)
+        {
+            delete this;
+        }
+
+        return count;
+    }
+
+
+    HRESULT STDMETHODCALLTYPE CorProfiler::Initialize(IUnknown* pICorProfilerInfoUnk)
+    {
+        //
+        // Get and set profiler pointers
+        //
+        if (m_dispatcher == nullptr)
+        {
+            return E_FAIL;
+        }
+        DynamicInstance* cpInstance = m_dispatcher->GetContinuousProfilerInstance();
+        if (cpInstance != nullptr)
+        {
+            m_cpProfiler = cpInstance->GetProfilerCallback();
+        }
+        DynamicInstance* tracerInstance = m_dispatcher->GetTracerInstance();
+        if (tracerInstance != nullptr)
+        {
+            m_tracerProfiler = tracerInstance->GetProfilerCallback();
+        }
+        DynamicInstance* customInstance = m_dispatcher->GetCustomInstance();
+        if (customInstance != nullptr)
+        {
+            m_customProfiler = customInstance->GetProfilerCallback();
+        }
+
+        //
+        // Get Profiler interface ICorProfilerInfo6 for net46+
+        //
+        ICorProfilerInfo6* info6 = nullptr;
+        HRESULT hr = pICorProfilerInfoUnk->QueryInterface(__uuidof(ICorProfilerInfo6), (void**) &info6);
+        if (FAILED(hr))
+        {
+            Warn("Failed to attach profiler: interface ICorProfilerInfo6 not found.");
+            return E_FAIL;
+        }
+
+        // Gets the initial value for the event mask
+        DWORD mask_low;
+        DWORD mask_hi;
+        hr = info6->GetEventMask2(&mask_low, &mask_hi);
+        if (FAILED(hr))
+        {
+            Warn("Error getting the event mask.");
+            return E_FAIL;
+        }
+
+        Debug("MaskLow: ", mask_low);
+        Debug("MaskHi : ", mask_hi);
+
+        //
+        // Continuous Profiler Initialization
+        //
+        if (m_cpProfiler != nullptr)
+        {
+            HRESULT localResult = m_cpProfiler->Initialize(pICorProfilerInfoUnk);
+            if (SUCCEEDED(localResult))
+            {
+                // let's get the event mask set by the CP.
+                DWORD local_mask_low;
+                DWORD local_mask_hi;
+                HRESULT hr = info6->GetEventMask2(&local_mask_low, &local_mask_hi);
+                if (SUCCEEDED(hr))
```

Comment:
the mask is shared between all CorProfilers, the `info6` is extracted from the same `pIcorProfilerInfoUnk`.

Reply:
Also each profiler sets their own mask overwriting the previous mask. So the logic here is to extract each local mask and do a bitwise OR to get the global one that satisfy every profiler.

Reply:
Ok, that's the thing, in 168/169 there are not the same thing. Let me explain it:

133 - We get the `ICorProfilerInfo5` interface from the `pICorProfilerInfoUnk` given by the runtime. Note: `pICorProfilerInfoUnk` is shared with CP and Tracer profilers, so their `ICorProfilerInfoX` will be extracted the same way we do here. And in fact, the mask is global and shared with all profilers.

144 - We read the Profiler mask using the `ICorProfilerInfo5` instance we got from 133.

159 - We call the `Initialize` function from the ContinousProfiler and we pass the same `pICorProfilerInfoUnk` instance. In this step the Continous Profiler will set the required event mask to work.

165 - We read again the event mask from our `ICorProfilerInfo5` instance, and because we are using the same `pICorProfilerInfoUnk` we will see here the event masks that the continuous profiler set.

168, 169 - We do the OR thing we our global masks.

Then we repeat the same thing with the other profilers... Each profiler will replace the existing mask with their masks, that's why we are doing the OR.

When we finish we manually sets the global mask in 257 and that change will affect the current runtime instance so it will affect all profilers.

Reply:
Thank you for explaining. We should add some comments to help future readers to understand this :)

Reply:
❗  [follow up in future PR]

Reply:
I've added a comment in 99f9c82.